### PR TITLE
BPU refactoring

### PR DESCRIPTION
**Pull request by dhschall**
This set of patches combines multiple enhancements in the branch predictor unit. Specifically it:
- Adds a new BranchType class useful for better statistics and debugging.
- Creates Separate SimObjects for BTB, RAS and Indirect branch predictor to allow different implementation of each.
- Moves the logics log RAS and indirect branch predictor management out of the main BPU to simplify its design.
- Improves statistics

The changes to the branch predictor are required to implement the decoupled front-end

### CODE REVIEW

**Comment by powerjg** [comment]
It would be great if @pcabre or @jairobalart are around and could provide feedback since they were the last ones to do a refactor of this code.

**Comment by Harshil2107** [comment]
There are 13 tests failing in the testlib-quick. 

All the failed tests fail with he following error message:

```
AttributeError: Class HPI_BP has no parameter 'BTBEntries'
```
The failing tests are:
- test-low_power-close_adaptive-ALL-x86_64-opt
- test-low_power-close_adaptive-ALL-x86_64-opt-MatchStdoutNoPerf
- test-low_power-open_adaptive-ALL-x86_64-opt
- test-low_power-open_adaptive-ALL-x86_64-opt-MatchStdoutNoPerf
- realview64-simple-atomic-ALL-x86_64-opt
- realview64-simple-atomic-dual-ALL-x86_64-opt
- realview64-simple-atomic-dual-ALL-x86_64-opt-MatchFileRegex
- realview64-simple-atomic-checkpoint-ALL-x86_64-opt
- realview64-simple-timing-ALL-x86_64-opt
- realview64-simple-timing-dual-ALL-x86_64-opt
- realview64-simple-timing-dual-ALL-x86_64-opt-MatchFileRegex
- realview64-switcheroo-atomic-ALL-x86_64-opt
- realview64-switcheroo-timing-ALL-x86_64-opt

I am pasting the complete stderr output from the test `test-low_power-close_adaptive-ALL-x86_64-opt` below:
```
AttributeError: Class HPI_BP has no parameter 'BTBEntries'

At:
  src/python/m5/SimObject.py(412): __setattr__
  src/python/m5/SimObject.py(297): __init__
  /__w/gem5/gem5/configs/common/cores/arm/HPI.py(1682): <module>
  <frozen importlib._bootstrap>(241): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(883): exec_module
  <frozen importlib._bootstrap>(703): _load_unlocked
  <frozen importlib._bootstrap>(1006): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1027): _find_and_load
  <frozen importlib._bootstrap>(1050): _gcd_import
  /usr/lib/python3.10/importlib/__init__.py(126): import_module
  /__w/gem5/gem5/configs/common/cores/arm/__init__.py(44): <module>
  <frozen importlib._bootstrap>(241): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(883): exec_module
  <frozen importlib._bootstrap>(703): _load_unlocked
  <frozen importlib._bootstrap>(1006): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1027): _find_and_load
  <frozen importlib._bootstrap>(1050): _gcd_import
  /usr/lib/python3.10/importlib/__init__.py(126): import_module
  /__w/gem5/gem5/configs/common/ObjectList.py(148): _add_objects
  /__w/gem5/gem5/configs/common/ObjectList.py(115): __init__
  /__w/gem5/gem5/configs/common/ObjectList.py(176): <module>
  <frozen importlib._bootstrap>(241): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(883): exec_module
  <frozen importlib._bootstrap>(703): _load_unlocked
  <frozen importlib._bootstrap>(1006): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1027): _find_and_load
  <frozen importlib._bootstrap>(241): _call_with_frames_removed
  <frozen importlib._bootstrap>(1079): _handle_fromlist
  /__w/gem5/gem5/configs/dram/low_power_sweep.py(45): <module>
  src/python/m5/main.py(642): main
```

File: src/cpu/pred/BranchPredictor.py [hunk 1]
```diff
@@ -63,14 +178,23 @@ class BranchPredictor(SimObject):
     abstract = True
 
     numThreads = Param.Unsigned(Parent.numThreads, "Number of threads")
-    BTBEntries = Param.Unsigned(4096, "Number of BTB entries")
```
Comment (andysan) [inline src/cpu/pred/BranchPredictor.py:66]: This will affect quite a few configuration files in `configs/common/cores/` and `src/python/gem5/prebuilt`. A quick `git grep` should find them.

Comment (dhschall) [reply src/cpu/pred/BranchPredictor.py:66]: I updated all config files where I could find the BTB parameters
Now all CI test pass

File: src/cpu/pred/BranchPredictor.py [hunk 2]
```diff
@@ -63,14 +178,23 @@ class BranchPredictor(SimObject):
...
     instShiftAmt = Param.Unsigned(2, "Number of bits to shift instructions by")
+    requiresBTBHit = Param.Bool(
+        False,
+        "Requires a BTB hit to detect if "
+        " a branch was a return or indirect branch.",
+    )
 
+    RASSize = Param.Unsigned(16, "RAS size")
+
+    BTB = Param.BranchTargetBuffer(SimpleBTB(), "Branch target buffer (BTB)")
```
Comment (andysan) [inline src/cpu/pred/BranchPredictor.py]: I'd suggest calling these `btb` and `ras` instead since we don't normally capitalise abbreviations on their own (see for example `mmu` in `BaseCPU`). It's likely that there will be modules with CPU configurations containing a BTB class called just `BTB`, so this would avoid a bit of confusion.

File: src/cpu/pred/bi_mode.cc [hunk 3]
```diff
@@ -221,7 +238,7 @@ BiModeBP::update(ThreadID tid, Addr branchAddr, bool taken, void *bpHistory,
         }
     }
 
-    delete history;
+    delete history; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/bi_mode.cc:241]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/bi_mode.cc [hunk 4]
```diff
@@ -82,16 +94,27 @@ BiModeBP::uncondBranch(ThreadID tid, Addr pc, void * &bpHistory)
...
+
+
+void
+BiModeBP::squash(ThreadID tid, void * &bpHistory)
 {
     BPHistory *history = static_cast<BPHistory*>(bpHistory);
     globalHistoryReg[tid] = history->globalHistoryReg;
 
-    delete history;
+    delete history; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/bi_mode.cc:117]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/ltage.cc [hunk 5]
```diff
@@ -132,19 +144,19 @@ LTAGE::update(ThreadID tid, Addr branch_pc, bool taken, void* bp_history,
     tage->updateHistories(tid, branch_pc, taken, bi->tageBranchInfo, false,
                           inst, corrTarget);
 
-    delete bi;
+    delete bi; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/ltage.cc:147]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/multiperspective_perceptron.cc [hunk 6]
```diff
@@ -812,21 +833,15 @@ MultiperspectivePerceptron::update(ThreadID tid, Addr instPC, bool taken,
...
     threadData[tid]->last_ghist_bit = taken;
 
-    delete bi;
-}
-
-void
-MultiperspectivePerceptron::btbUpdate(ThreadID tid, Addr branch_pc,
-                                      void* &bp_history)
-{
+    delete bi; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/multiperspective_perceptron.cc:836]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/multiperspective_perceptron.cc [hunk 7]
```diff
@@ -812,21 +833,15 @@ MultiperspectivePerceptron::update(ThreadID tid, Addr instPC, bool taken,
...
 void
-MultiperspectivePerceptron::squash(ThreadID tid, void *bp_history)
+MultiperspectivePerceptron::squash(ThreadID tid, void * &bpHistory)
 {
-    assert(bp_history);
-    MPPBranchInfo *bi = static_cast<MPPBranchInfo*>(bp_history);
-    delete bi;
+    assert(bpHistory);
+    MPPBranchInfo *bi = static_cast<MPPBranchInfo*>(bpHistory);
+    delete bi; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/multiperspective_perceptron.cc:844]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/multiperspective_perceptron_tage.cc [hunk 8]
```diff
@@ -664,25 +676,32 @@ MultiperspectivePerceptronTAGE::update(ThreadID tid, Addr instPC, bool taken,
                                   false, inst, corrTarget);
         }
     }
-    delete bi;
+    delete bi; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/multiperspective_perceptron_tage.cc:679]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/simple_indirect.cc [hunk 9]
```diff
@@ -64,174 +83,310 @@ SimpleIndirectPredictor::SimpleIndirectPredictor(
...
+}
+
+
+
 inline Addr
-SimpleIndirectPredictor::getSetIndex(Addr br_addr, unsigned ghr, ThreadID tid)
+SimpleIndirectPredictor::getSetIndex(Addr br_addr, ThreadID tid)
 {
-    ThreadInfo &t_info = threadInfo[tid];
+    ThreadInfo &ti = threadInfo[tid];
```
Comment (Harshil2107) [inline src/cpu/pred/simple_indirect.cc:374]: Let us change this back to `t_info`, it seems more explicit than `ti`.

File: src/cpu/pred/tage.cc [hunk 10]
```diff
@@ -83,15 +95,15 @@ TAGE::update(ThreadID tid, Addr branch_pc, bool taken, void* bp_history,
     // optional non speculative update of the histories
     tage->updateHistories(tid, branch_pc, taken, tage_bi, false, inst,
                           corrTarget);
-    delete bi;
+    delete bi; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/tage.cc:98]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/tage.cc [hunk 11]
```diff
@@ -83,15 +95,15 @@ TAGE::update(ThreadID tid, Addr branch_pc, bool taken, void* bp_history,
...
 
 void
-TAGE::squash(ThreadID tid, void *bp_history)
+TAGE::squash(ThreadID tid, void * &bpHistory)
 {
-    TageBranchInfo *bi = static_cast<TageBranchInfo*>(bp_history);
+    TageBranchInfo *bi = static_cast<TageBranchInfo*>(bpHistory);
     DPRINTF(Tage, "Deleting branch info: %lx\n", bi->tageBranchInfo->branchPC);
-    delete bi;
+    delete bi; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/tage.cc:106]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/tage_sc_l.cc [hunk 12]
```diff
@@ -462,7 +474,7 @@ TAGE_SC_L::update(ThreadID tid, Addr branch_pc, bool taken, void *bp_history,
                               inst, corrTarget);
     }
 
-    delete bi;
+    delete bi; bpHistory = nullptr;
```
Comment (Harshil2107) [inline src/cpu/pred/tage_sc_l.cc:477]: Let's move `bpHistory = nullptr;` to a new line.

File: src/cpu/pred/2bit_local.cc [hunk 13]
```diff
@@ -67,15 +79,15 @@ LocalBP::LocalBP(const LocalBPParams &params)
...
 {
-// Place holder for a function that is called to update predictor history when
-// a BTB entry is invalid or not found.
+// Place holder for a function that is called to update predictor history
 }
 
 
 bool
-LocalBP::lookup(ThreadID tid, Addr branch_addr, void * &bp_history)
+LocalBP::lookup(ThreadID tid, Addr branch_addr, void * &bpHistory)
```
Comment (Harshil2107) [inline src/cpu/pred/2bit_local.cc:90]: Can we revert the changes from `bpHistory` back to `bp_history` in this and all the files where this change is made?

I think `bp_history` better conforms to the gem5 style guide: https://www.gem5.org/documentation/general_docs/development/coding_style/

**Review by Harshil2107 — CHANGES_REQUESTED** [review]

**Comment by dhschall** [comment]
Thank you @Harshil2107 for the comments.
I talked with @andysan that all of it is too much in one change. Also, the PR includes both refactoring and new functionality.
I'll split this PR into several to make the review process easier. Of course, I'll still incorporate all your suggestions.

The first one is #412
